[Profiling] Add TopN functions API

jbcrail opened the pull request:
This PR adds a new API call `/_profiling/topn/functions` that builds on top of the existing `/_profiling/stacktraces` API, but moves all computation for TopN functions from Kibana to Elasticsearch.

**Note**: this PR is currently in draft and is incomplete (awaiting serialization/deserialization for nested objects).

At least two Kibana PRs are required to integrate this new API (https://github.com/elastic/kibana/pull/171151 is already pending).

jbcrail committed:
Add stub for TopN functions endpoint

jbcrail committed:
Remove whitespace

jbcrail committed:
Add type for stackframe metadata

jbcrail committed:
Add implementation

danielmitterdorfer reviewed (commented):
I had a brief look at this (i.e. no thorough review yet) and left a few comments in code but I have a couple of high-level comments:

* Given that the UI shows only a limited number of functions (100), can we introduce a request parameter to limit the results to that number? This reduces serialization time. It would probably also make sense to introduce a separate request class here.
* I sense we could also take advantage of the fact that we only need a limited number of functions by adapting `TransportGetStackTracesAction`. That class has become quite large already and we should first refactor it and break it into smaller pieces (separate PR).

Did you get a chance to measure end to end response time (i.e. how much does this PR improve things)?

jbcrail committed:
Simplify listener

jbcrail committed:
Return Top N metadata

elasticsearchmachine commented:
Pinging @elastic/profiling (Team:Universal Profiling)

jbcrail commented on x-pack/plugin/profiling/src/main/java/org/elasticsearch/xpack/profiling/TransportGetTopNFunctionsAction.java:
```diff
@@ -0,0 +1,183 @@
+    protected void doExecute(Task task, GetStackTracesRequest request, ActionListener<GetTopNFunctionsResponse> listener) {
+        Client client = new ParentTaskAssigningClient(this.nodeClient, transportService.getLocalNode(), task);
+        StopWatch watch = new StopWatch("getTopNFunctionsAction");
+        client.execute(GetStackTracesAction.INSTANCE, request, new ActionListener<>() {
```
Thanks, I found a good example in `TransportGetStackTracesAction`.

jbcrail commented on x-pack/plugin/profiling/src/main/java/org/elasticsearch/xpack/profiling/TransportGetTopNFunctionsAction.java:
```diff
@@ -0,0 +1,183 @@
+        client.execute(GetStackTracesAction.INSTANCE, request, new ActionListener<>() {
+            @Override
+            public void onResponse(GetStackTracesResponse response) {
+                long responseStart = System.nanoTime();
```
It was copied from `TransportGetFlamegraphAction` originally. Since the flamegraph and TopN functions use the stacktrace data and have a similar algorithm, I started from `TransportGetFlamegraphAction` when working on this PR.

It looks like we could clean up `TransportGetFlamegraphAction.doExecute` in the same way as suggested below.

jbcrail commented:
Thanks for the initial pass, @danielmitterdorfer.

As for limiting the returned functions, the existing Kibana API has a limit parameter, but the UI effectively ignores it by requesting 100k functions (a maximum not typically reached for most time ranges). The TopN function table in Kibana sorts the records in place without calling back to ES whenever a user changes the sort order on any column. I agree adding a limit parameter (along with a way to override the default sort) to the new API would have much better performance, but we would have to coordinate with Kibana. I chose to simplify this API endpoint initially, so that we would get a large performance boost now and further improvements in the future.

I also agree that there are many refactoring opportunities that we can follow up soon after.

I did run a few comparisons between:
1. the existing Kibana API endpoints for the flamegraph and TopN functions, and
2. the Elasticsearch API endpoints for the flamegraph and TopN functions.

For the first comparison, the TopN function response time is 1.5-2x slower than the flamegraph response time.

For the second comparison, it appears anecdotally the response time for both endpoints is almost equal. Once the new endpoint is connected up in Kibana, I would expect the slower performance noted in the first comparison would go away.

danielmitterdorfer commented on x-pack/plugin/profiling/src/main/java/org/elasticsearch/xpack/profiling/TransportGetTopNFunctionsAction.java:
```diff
@@ -0,0 +1,171 @@
+
+                for (Frame frame : stackFrame.frames()) {
+                    String frameGroupId = FrameGroupID.create(fileId, addressOrLine, executable, frame.fileName(), frame.functionName());
+                    if (builder.setCurrentTopNFunction(frameGroupId) == false) {
```
It's confusing to me that a setter returns a boolean and I constantly need to look at the implementation to understand what it means. Can we change the implementation so it's more apparent what's happening. E.g. if I look at similar code in `TransportGetFlamegraphAction` I have a much better grasp of what's happening on a high level.

danielmitterdorfer commented on x-pack/plugin/profiling/src/main/java/org/elasticsearch/xpack/profiling/TransportGetTopNFunctionsAction.java:
```diff
@@ -0,0 +1,171 @@
+                            )
+                        );
+                    }
+                    if (frameGroupsPerStackTrace.contains(frameGroupId) == false) {
```
I *think* I understand why this guard is here but IMHO this warrants a longer comment explaining it (I believe it's there because we want to avoid counting recursive function calls multiple times in the same stack trace)

jbcrail replied:
You're right about the reason behind the guard. I added comments to clarify.

danielmitterdorfer commented on x-pack/plugin/profiling/src/main/java/org/elasticsearch/xpack/profiling/TransportGetTopNFunctionsAction.java:
```diff
@@ -0,0 +1,171 @@
+                        frameGroupsPerStackTrace.add(frameGroupId);
+                        builder.addToCurrentInclusiveCount(samples);
+                    }
+                    if (i == frameCount - 1) {
```
Can we mention that this is a leaf frame?

danielmitterdorfer commented on x-pack/plugin/profiling/src/main/java/org/elasticsearch/xpack/profiling/TransportGetTopNFunctionsAction.java:
```diff
@@ -0,0 +1,171 @@
+            Collections.sort(functions, Collections.reverseOrder());
+            long sumSelfCPU = 0;
+            long sumTotalCPU = 0;
+            for (int i = 0; i < functions.size(); i++) {
```
Can we use regular iteration here and have a separate `rank` counter that we increase in the body?

danielmitterdorfer reviewed (commented):
Thanks! I left a few more comments. Contrary to the flamegraph API that we've introduced recently I do believe that we should finalize the API between Kibana and Elasticsearch before we release it (I don't see flamegraph rendering changing significantly but I can see that we change the view model for topN functions). One important reason is that this is a new API and once it is introduced we're bound by backwards compatibility guarantees that will hinder progress (assuming we'd introduce this in 8.12.0, we need to ensure that an upgrade from 8.12.0 to 8.16.0 works smoothly, i.e. an 8.12.0 Kibana must be able to call this API on an 8.16.0 ES cluster during the upgrade procedure). I suppose Kibana pulls data for all functions because we need to sort them on the fly by different properties (self CPU, total CPU, CO2 usage etc.) but IMHO we should strive to avoid serializing up to 1000 times more data than we actually need.

jbcrail committed:
Add comments

jbcrail committed:
Remove template type

jbcrail committed:
Remove unused variable

jbcrail commented on x-pack/plugin/profiling/src/main/java/org/elasticsearch/xpack/profiling/TransportGetTopNFunctionsAction.java:
```diff
@@ -52,25 +52,13 @@ public TransportGetTopNFunctionsAction(NodeClient nodeClient, TransportService t
-            }
-        });
+        client.execute(GetStackTracesAction.INSTANCE, request, ActionListener.wrap(searchResponse -> {
+            long responseStart = System.nanoTime();
```
Thanks, I missed that.

danielmitterdorfer commented on x-pack/plugin/profiling/src/main/java/org/elasticsearch/xpack/profiling/RestGetTopNFunctionsAction.java:
```diff
@@ -0,0 +1,45 @@
+
+import static org.elasticsearch.rest.RestRequest.Method.POST;
+
+public class RestGetTopNFunctionsAction extends BaseRestHandler {
```
Can you please annotate this with `@ServerlessScope(Scope.PUBLIC)`? This is required so the API can be called at all from Kibana in Serverless.

elasticsearchmachine commented:
Pinging @elastic/obs-knowledge-team (Team:obs-knowledge)

jbcrail committed:
Add serverless scope

jbcrail committed:
Update how action is performed

This was needed to align with recent flamegraph improvements.

jbcrail commented:
I moved this PR back to draft since enough drift has happened since the end of last year.

The PR will move back for review once I have addressed the subsequent drift and also addressed key design decisions identified by the initial feedback.

The design decisions that I need to address are:
- flatten the response into a column-oriented format (similar to flamegraph) for reduced payload and improved response time
- add configurable parameters for pagination and sorting

danielmitterdorfer commented:
@jbcrail I'm going to close this PR as I'm working on a revised version of this API for our APM integration.
